packages/core/src/application.ts

```diff
@@ -502,6 +504,105 @@ export class Application extends Context {
     });
   }
 
+  async start(): Promise<void> {
+    await Promise.all([
+      this._startHttp(),
+      this._startGrpc(),
+    ]);
+  }
+
+  async _startGrpc(): Promise<void> {
+    debug('Setting up gRPC server');
+    const server = new grpc.Server();
+
+    for (const b of this.find('controllers.*')) {
+      const controllerName = b.key.replace(/^controllers\./, '');
+      const ctor = b.valueConstructor;
+      if (!ctor) {
+        throw new Error(
+          `The controller ${controllerName} was not bound via .toClass()`);
+      }
+
+      // tslint:disable-next-line:no-any
+      const spec: any = (ctor as {grpcService?: object}).grpcService;
+      if (!spec) {
+        debug(
+          `  skipping controller ${controllerName} - no gRPC API was specified`,
+        );
+        continue;
+      }
+
+      debug('  adding controller %s with spec %j', controllerName, spec);
+      for (const key in spec) {
+        const methodName = key;
+        const opSpec = spec[key];
+        // FIXME(bajtos) handle the case when the controller method is using
+        // opSpec.originalName
+
+        if (!ctor.prototype[methodName]) {
+          debug(
+            '   - skipping method %s - %s.%s not implemented',
+            opSpec.originalName,
+            controllerName,
+            methodName);
+          continue;
+        }
+
+        debug(
+          '   - added %s as %s.%s',
+          opSpec.originalName,
+          controllerName,
+          methodName,
+        );
+
+        // TODO: support stream method types
+        server.register(
+          opSpec.path,
+          createHandlerFor(controllerName, methodName, this),
+          opSpec.responseSerialize,
+          opSpec.requestDeserialize,
+          'unary',
+        );
+
+      }
+    }
+
+    function createHandlerFor(
+      controllerName: string,
+      methodName: string,
+      rootContext: Context,
+    ) {
+      // TODO(bajtos) support metadata?
```

Comment:
`request.metadata` contains gRPC metadata from the client call, like http headers.